source/language/directives.rst

```diff
@@ -1,14 +1,95 @@
 Directives
 ==========
 
-OpenQASM supports a directive mechanism that allows other information to
-be included in the program. A directive begins with ``#pragma`` and
-terminates at the end of the line. Directives can provide annotations
-that give additional information to compiler passes and the target
-system or simulator. Ideally the meaning of the program does not change
+OpenQASM supports directive mechanisms that allows other information to
+be included in the program. Directives are either pragmas or annotations.
+Both are used to supply additional information to the compiler passes and the
+target system or simulator. Ideally the meaning of the program does not change
 if some or all of the directives are ignored, so they can be interpreted
 at the discretion of the consuming process.
 
+Pragmas
+-------
+
+Pragma directives start with ``pragma`` and continue to the end of line. The
+text after ``pragma`` is a single string, and parsing is left to the specific
+implementation. The ``#pragma`` syntax is deprecated, but is accepted for
+compatibility. Pragmas should be processed as soon as they are encountered; if a
+pragma is not supported by a compiler pass it should be ignored and preserved
+intact for future passes.  Ideally, pragmas should be considered global and
+immutable; avoiding stateful interactions is recommended to prevent unexpected
+behaviors between source files.
+
+Pragmas are useful for extending OpenQASM functionality that is not described in
+this specification, such as adding noise operations.
+
+\*Note: The following examples are simply possible implementations, this
+specification does not define any pragmas. Please consult your tool's
+documentation for supported pragmas.
+
+.. code-block:: c
+   :force:
+
+   rx(0.15) q;
+   pragma noise bit_flip(0.1)
+
+Pragmas can also be used to specify system-level information or assertions for
+the entire circuit.
+
+.. code-block:: c
+   :force:
+
+   OPENQASM 3.0;
+
+   // Attach billing information
+   pragma user alice account 12345678
+
+   // Assert that the QPU is healthy to run this circuit
+   pragma max_temp qpu 0.4
+
+   qubit[2] q;
+
+
+Annotations
+-----------
+
+Annotations can be added to supply additional information to the following
+OpenQASM statement. These annotations will apply to the subsequent declaration,
+box, or single line of code. Annotations will start with a ``@`` symbol, have an
```

Comment:
By "single line of code", do we mean that I can do:
```qasm
@storage fast
int[8] a1 =
@overflow check
100 +
@overflow nocheck
100;
```
(i.e. deliberately splitting the line to apply different annotations to different subexpressions)?

I'd be in favour of annotations being applied to the following "statement" rather than line (maybe this was what was intended already). I can only think of grossly contrived examples for splitting within statements, and implementation could be very tricky if annotations can be attached to arbitrary groups of tokens rather than fixed parser productions.

Applying to the following "statement" is potentially awkward if somebody decides to follow an annotation with several statements on the same line:
```qasm
@annotation
int[8] a = 1; uint[8] b = 2;
```
so it's not necessarily clear cut.  I'd prefer to say that this example only applied the annotation to `int[8] a = 1;` and push the burden of complaining about the misleading formatting onto a linter.

Reply:
The intention was "statement" in terms of the grammatical definition. I over-elaborated because I wasn't necessarily certain that a casual reader would understand the word "statement" in context, but it may be clearer to not over explain (and maybe provide a link to a definition for "statement").

Reply:
Perfect, thanks - that makes more sense to me.

Reply:
This comment chain is completely addressed by da1799d.